src/meta/src/barrier/context/recovery.rs

```diff
@@ -508,20 +502,28 @@ impl GlobalBarrierWorkerContextImpl {
                         .await?;
 
                     let background_jobs = {
-                        let jobs = self
+                        let mut unfinished_jobs = HashMap::new();
+                        let mut background_jobs = self
                             .list_background_job_progress()
                             .await
                             .context("recover background job progress should not fail")?;
-                        let mut background_jobs = HashMap::new();
-                        for (definition, stream_job_fragments) in jobs {
-                            background_jobs
-                                .try_insert(
-                                    stream_job_fragments.stream_job_id(),
-                                    (definition, stream_job_fragments),
-                                )
-                                .expect("non-duplicate");
+                        for (job_id, job_info) in info.values().flatten() {
+                            let Some(definition) = background_jobs.remove(job_id) else {
+                                continue;
+                            };
+                            if job_info.tracking_progress_actor_ids().is_empty() {
+                                // If there's no tracking actor in the job, we can finish the job directly.
+                                self.metadata_manager
+                                    .catalog_controller
+                                    .finish_streaming_job(job_id.table_id as _)
+                                    .await?;
```

Comment:
Noticed that we moved `finish_stream_job` path here, from before `pre_apply_drop_cancel`, and actor scaling. Is that intentional? Not sure if there will be any correctness issues.